feat(repo): Add `grz-watchdog`

tedil opened the pull request:
(this is currently based on `origin/upgrade-pydantic`)
This PR adds `grz-watchdog`, a snakemake workflow for automating the ingest/submission process on the GRZ side.
Can be run both for single submissions and in batch mode (a continuous monitoring mode may follow later, using snakemake's `from_queue`). See README.md for more details. 

~Also comes with a docker file for containerization, as well as a docker-compose definition for setting up a local minio + mock bfarm api for testing purposes.~
There is now a `build.sh` that constructs and then builds a (multi-stage) Dockerfile from `snakemake --containerize` and the `Dockerfile.template` (which contains instructions to install pixi and grz-watchdog).

The docker image uses `packages/grz-watchdog/container_utils/scripts/entrypoint.sh` as its entrypoint (which basically is `pixi run`) and has two convenience scripts:
- `process-single <submitter_id> <inbox> <submission_id> <qc_target> [snakemake_options...]` which processes a single known submission
- `process-batch [snakemake_options…]` which scans all configured inboxes and submits multiple detected submissions to be processed by snakemake

feat(grz-watchdog): Add `grz-watchdog`

ci(repo): include grz-watchdog in list of allowed scopes

ci(repo): perform snakefmt and snakemake lint in checks for grz-watchdog

chore(repo): update uv.lock

chore(repo): adapt to new lints

tedil committed:
move deprecated model config serializers to custom ones

# Conflicts:
#	packages/grz-db/pyproject.toml

tedil committed:
move deprecated model config serializers to custom ones

tedil committed:
move annotated after validators to field validators

tedil committed:
move annotated after validators to field validators

tedil committed:
remove id from SubmissionBase

tedil committed:
remove id from SubmissionBase

tedil committed:
move submission id validator to field_validator

tedil committed:
move submission id validator to field_validator

tedil committed:
add extend_existing to table_args for now, probably a workaround

tedil committed:
add extend_existing to table_args for now, probably a workaround

tedil committed:
use model_fields on class, not instance

tedil committed:
use model_fields on class, not instance

tedil committed:
bump pydantic version

tedil committed:
bump pydantic version

tedil committed:
update uv.lock

tedil committed:
potentially deal with pydantic/sqlalchemy changes regarding class type fields

tedil committed:
potentially deal with pydantic/sqlalchemy changes regarding class type fields

tedil committed:
ignore COM812

tedil committed:
ignore COM812

tedil committed:
try clearing sqlmodel metadata automatically (doesn't seem to work)

tedil committed:
try clearing sqlmodel metadata automatically (doesn't seem to work)

tedil committed:
fix lints

tedil committed:
fix pydantic ClassVar issues

tedil committed:
fix pydantic ClassVar issues

tedil committed:
fix pydantic ClassVar issues, part2

tedil committed:
fix pydantic ClassVar issues, part2

tedil committed:
update uv.lock

tedil committed:
grz-cli: update uv.lock

tedil committed:
grzctl: update uv.lock

tedil committed:
grz-common: update uv.lock

tedil committed:
grz-db: update uv.lock

tedil committed:
Merge branch 'main' into upgrade-pydantic

tedil committed:
update uv.locks for pydantic changes

tedil committed:
fix imports

tedil committed:
ClassVar cannot contain any type vars

Hoeze commented on packages/grz-watchdog/tests/test_watchdog_single.py:
```diff
@@ -0,0 +1,188 @@
+        container_temp_path = f"/tmp/{Path(encrypted_file_key).name}"
+
+        run_in_container(*PIXI_RUN_PREFIX, "mc", "cp", s3_path_to_corrupt, container_temp_path)
+        corruption_command = f"printf '\\1' | dd of={container_temp_path} bs=1 count=1 conv=notrunc"
```
that `dd` makes me nervous :smile:

twrightsman replied:
Could corrupt purely in Python

tedil committed:
feat(grz-watchdog): add grz-watchdog workflow

tedil committed:
test(grz-watchdog): add testing setup

tedil committed:
chore(repo,grz-watchdog): include grz-watchdog in ci

tedil committed:
chore(repo): update pyproject.toml

tedil committed:
chore(repo): include grz-watchdog in release-please-config

tedil committed:
grouped remaining changes

tedil committed:
merge origin/upgrade-pydantic

tedil committed:
feat(grz-watchdog): add container setup

tedil committed:
feat(grz-watchdog): add test container setup

tedil committed:
add outdated documentation

tedil committed:
add .gitignore

tedil committed:
reinstate accidentally deleted file

tedil committed:
add .containerignore

tedil committed:
test(grz-db): Add get monthly qc stats grz-db test

tedil committed:
un-conda-fy workflow, update Dockerfile, use pixi

tedil committed:
update pixi.toml & pixi.lock

tedil committed:
update containter util scripts / entrypoint

tedil committed:
add corresponding pixi tasks

tedil committed:
cleanup & unify pixi and container setup

tedil committed:
add runner.py

tedil committed:
… remembered script name incorrectly

tedil committed:
default to grz-watchdog dir as workdir

tedil committed:
add mock_api Dockerfile

tedil committed:
add speed estimates to config+schema

tedil committed:
update pixi to 0.59.0 in Dockerfile

tedil committed:
make daemon rules local

tedil committed:
chore(repo): update pixi version in ci

tedil committed:
use pixi 0.58.0

tedil committed:
chore(repo): use pixi 0.58.0 in ci

tedil committed:
add missing metadata to rule archive inputs

tedil committed:
add missing metadata to rule qc inputs

tedil committed:
do not include 'uploading' submissions in the list of available_submission

tedil committed:
fix runtime estimates

tedil committed:
create temporary configfiles

tedil committed:
Merge branch 'add-grz-watchdog' of github.com:bfarm-mvh/grz-tools into add-grz-watchdog

tedil committed:
move qc run shell code into script

tedil committed:
simplify run_qc script + params etc

tedil committed:
also cleanup prepare qc shell → script

tedil committed:
update README.md

tedil committed:
fix prepare_qc script

tedil committed:
fix testing quoting

tedil committed:
.

tedil committed:
fix print

tedil committed:
update README.md with info about snakemake resources

tedil committed:
allow downloading references instead of building from scratch

tedil committed:
shift qc mem requirements to estimate_qc_memory function in common.smk

tedil committed:
do not reference prepare_qc_workflow_references rule directly

tedil committed:
merge origin/main

tedil committed:
do not limit grzctl list output

tedil committed:
do not limit grzctl list output, part 2

tedil committed:
'uploaded' means available

tedil committed:
populate detailed qc

tedil committed:
populate basic qc

tedil committed:
lowercase db_config…

tedil committed:
double quote

tedil committed:
fix perhaps temp

tedil committed:
don't register qc results twice

tedil committed:
allow continuing from other reasonable '*ed' states

tedil committed:
Merge branch 'add-grz-watchdog' of github.com:bfarm-mvh/grz-tools into add-grz-watchdog

tedil committed:
add continuable states

tedil committed:
do not depend on db sync in metadata and process_submission rule

tedil committed:
refactor(grz-cli,grz-common,grzctl): have command accept each worker dir separately/fine grained control

tedil committed:
re-init db for every test (so workflow doesn't require dependency on init_db rule all the time)

tedil committed:
add test for resuming from intermediate files

tedil committed:
refactor(grz-watchdog): adapt to grz-cli/grzctl cli changes, refactor workflow accordingly, no more shared mutable submission directory

tedil committed:
add missing check_and_register script

tedil committed:
remove unnecessary dependency from process_submission

tedil committed:
create symlink dir structure for nextflow

tedil committed:
rename to logs_dir

tedil committed:
rename to bundled/granular mode

tedil committed:
explicit in and outputs in snakemake workflow

tedil committed:
rename to bundled/granular mode, part2

tedil committed:
fix unbound vars

tedil committed:
fix more unbound vars; make archive also granular

tedil committed:
ensure timestamps are utc for sorting

tedil committed:
use temp,remove perhaps_temp, point to --notemp, fmt

tedil committed:
test_resumability_after_decrypt_interrupt: use --until instead of concrete intermediate target to keep temp files

tedil committed:
explicitly add coin-or-cbc to dependencies

tedil committed:
include decrypt outputs in get_successful_finalize_inputs for better resumability

tedil committed:
adjust interruption tests

tedil committed:
merge origin/main

tedil committed:
instead of interrupt, corrupt some file to ensure qc fails

tedil committed:
fmt

tedil committed:
sigint, then sigkill

tedil committed:
don't care about case in state checks

tedil committed:
rewrite interrupt test completely

tedil committed:
un-temp certain outputs

tedil committed:
add rule priorities for better scheduling

tedil committed:
fmt

tedil committed:
add slowifying nextflow config suitable for testing

tedil committed:
more tests for resumability

tedil committed:
allow configuring nextflow tmpdir

tedil committed:
Merge remote-tracking branch 'origin/main' into add-grz-watchdog

tedil committed:
Merge branch 'main' into add-grz-watchdog

tedil committed:
fix lints

tedil committed:
remove superfluous whitespace

tedil committed:
add grz_watchdog_changed to outputs explicitly

tedil committed:
fix expansion

tedil committed:
install podman from conda-forge instead

tedil committed:
don't run pixi install for non-existent pixi.toml

tedil committed:
remove podman ubuntu apparmor profile in ci

tedil committed:
try adding a custom apparmor profile for pixi installed podman instead

tedil committed:
disable apparmor altogether

tedil committed:
nuke

tedil committed:
try using a different cgroupmgr for podman?

tedil committed:
try installing newer podman from 25.04

tedil committed:
.

tedil committed:
put sudo in the correct location

tedil committed:
try more stuff

tedil committed:
smarter re-run logic

tedil committed:
psycopg lock

tedil committed:
don't re-register submissions

tedil committed:
only check validate checkpoint when needed

twrightsman commented on .github/workflows/ci.yml:
```diff
@@ -328,8 +338,113 @@ jobs:
+        run: |
+          cd ${{ matrix.package.path }}
+          pixi install --frozen
+          pixi run snakemake --lint    # TODO: check output and error if there are recommendations left
```
TODO for later?

tedil committed:
separate validation rule from decision checkpoint

twrightsman commented on packages/grz-db/src/grz_db/models/submission.py:
```diff
@@ -102,7 +103,7 @@ class SubmissionBase(SQLModel):
     immutable_fields: ClassVar[set[str]] = {"id"}
 
-    id: str
+    # id: str
```
Remove? I forget what the pydantic bump PR did

twrightsman reviewed (dismissed):
Only minor comments

tedil committed:
try working around snakemake metadata checking a little bit

tedil committed:
whoops, forgot to revert

tedil committed:
handle actually missing db state…

tedil commented on tests/conftest.py:
```diff
@@ -29,6 +31,11 @@
 
 
+@pytest.fixture(autouse=True)
+def clear_sqlmodel_metadata():
```
probably a child of the rebase + merge chaos in this pr (sorry).

tedil committed:
Merge remote-tracking branch 'origin/main' into add-grz-watchdog

tedil committed:
bump pixi version

tedil committed:
exclude code from rerun-triggers by default

tedil committed:
exclude pixi.lock from git for now

twrightsman commented on packages/grz-watchdog/workflow/rules/common.smk:
```diff
@@ -163,6 +163,44 @@ def get_validation_state(wildcards):
+    """
+    If the 'target_pattern' file exists, returns empty list [] (input not needed).
+    Otherwise, returns 'source'.
+    This prevents Snakemake from erroring on missing temp files if the job is done.
```
Just for my own understanding, why would it matter if any temp files are missing if the submission is `Finished`? Unless you mean resuming a submission where previous steps had their temp files deleted but they are no longer needed.

tedil committed:
Revise configuration options and error handling in README

Updated README.md to reflect changes in configuration options and error handling.

tedil committed:
Document executor options for grz-watchdog

Added information on running with different executors, including examples for slurm and links to executor plugins.

tedil committed:
don't set cleaning twice

tedil committed:
Merge remote-tracking branch 'origin/main' into add-grz-watchdog

tedil committed:
remove input from rerun triggers for default workflow profile

tedil committed:
introduce db_handles resource for rules writing to the db

tedil committed:
ensure db_state always is a string in check_and_register

tedil committed:
don't use shared launchdir for now

tedil committed:
allow checking out commit sha as well

tedil committed:
fix monthly qc stats for pg

tedil committed:
Merge remote-tracking branch 'origin/main' into add-grz-watchdog

tedil committed:
merge origin/main

use should_qc from db

tedil committed:
re-enable handover

tedil committed:
use timestamp of finished upload for pruefbericht generation only

tedil committed:
use rawfile instead

tedil committed:
db init marker should be considered ancient

tedil committed:
make check_and_register more robust

tedil committed:
fix logging in check_and_register

tedil committed:
remove unused imports

tedil committed:
merge origin/main

tedil committed:
merge origin/main

tedil committed:
update watchdog to changes from main

tedil committed:
snakefmt

tedil committed:
fix import

tedil committed:
actually do pixi env setup via uv because pixi doesn't work nicely with grz-check's maturin

tedil committed:
fix snakemake iocontainers changes

tedil committed:
Merge remote-tracking branch 'origin/main' into add-grz-watchdog

tedil committed:
merge origin/main

tedil committed:
merge origin/main

tedil committed:
merge origin/main

tedil committed:
re-enable relative pypi deps

tedil committed:
disable pixi pypi deps again, use setup task instead

tedil committed:
add snakemake-logger-plugin-prometheus to deps

tedil committed:
use prometheus plugin from pypi instead of bioconda for now

tedil committed:
track pixi.lock in repo

tedil committed:
merge origin/main

tedil committed:
bump prometheus logger again

tedil committed:
include db config during clean

tedil committed:
adapt to sop updates

tedil committed:
also anchor metadata for consent rule

tedil committed:
merge origin/main

tedil committed:
adapt to qc-workflow-version changes

tedil committed:
merge origin/main into add-grz-watchdog

tedil committed:
switch to unified config

tedil committed:
no need for dynamic cfg_path anymore, snakefmt

tedil committed:
use from_path

tedil committed:
include submitter-id in scan_inbox_and_augment

tedil committed:
add missing --inbox and --submitter-id flags where needed

tedil committed:
update snakefmt

tedil committed:
re-introduce cli modifications to support granular mode

tedil committed:
update dependency versions

tedil committed:
defensively always create in/output directories in shell scripts

tedil committed:
sync uv version to main project version

tedil committed:
fix wrapper arg forwarding

tedil committed:
have grzctl versions instead of wrapping grz-cli
